### PR TITLE
Add option to control unwind lowering

### PR DESCRIPTION
**Pull request by 1tnguyen**
- Added configurable AOT unwind-mode: cfg (default), dataflow, or none.
 
- Wired the option through target YAML, C++ `nvq++`, and Python compilation pipelines.
  
- Updated aggressive inlining pipeline which also runs unwind lowering

- Initialized the dataflow pass’s `dfJump` state (`None`) to prevent undefined value.

- Added C++/Python fake-target and focused MLIR regression coverage.

### CODE REVIEW

File: cudaq/include/cudaq/Optimizer/Transforms/Passes.h [hunk 1]
```diff
@@ -26,7 +26,8 @@ namespace cudaq::opt {
 /// Add a pass pipeline to transform call between kernels to direct calls that
 /// do not go through the runtime layers, inline all calls, and detect if calls
 /// to kernels remain in the fully inlined into entry point kernel.
-void addAggressiveInlining(mlir::OpPassManager &pm, bool fatalCheck = false);
+void addAggressiveInlining(mlir::OpPassManager &pm, bool fatalCheck = false,
+                           bool lowerUnwind = true);
```
Comment (schweitzpgi) [inline cudaq/include/cudaq/Optimizer/Transforms/Passes.h:30]: These are separate passes that ought to have no relationship to each other. I'm not sure why unwinding was buried in the inlining pipeline. It should be pulled out and stand on its own feet at this point.

File: cudaq/lib/Optimizer/Transforms/AggressiveInlining.cpp [hunk 2]
```diff
@@ -160,16 +160,18 @@ class CheckKernelCalls
...
 /// 3) Aggressively inline all calls.
 /// 4) Detect if kernel inlining has failed and left behind calls to kernels.
 /// Such a failure is most likely a sign that there is a cycle in the call
 /// graph. [This check is a bad idea: this should be deferred to final codegen
 /// when translating the final Quake IR.]
-void cudaq::opt::addAggressiveInlining(OpPassManager &pm, bool fatalChecks) {
+void cudaq::opt::addAggressiveInlining(OpPassManager &pm, bool fatalChecks,
+                                       bool lowerUnwind) {
   llvm::StringMap<OpPassManager> opPipelines;
-  pm.addNestedPass<func::FuncOp>(cudaq::opt::createUnwindLowering());
```
Comment (schweitzpgi) [inline cudaq/lib/Optimizer/Transforms/AggressiveInlining.cpp:172]: Yeah, this is truly weird. Unwinding ought to happen very early. Once done, it should never be thought about again. It's a pass for lowering very high-level syntactic goop from the front-end to a correct and selectable lower-level IR. That selectability is precisely what we want now.

File: cudaq/lib/Optimizer/Transforms/Pipelines.cpp
Comment (schweitzpgi) [inline cudaq/lib/Optimizer/Transforms/Pipelines.cpp]: While we're here, can you align this name with the actual name so as to reduce confusion? `"python-aot-pipeline"` would be consistent with what this actually is rather than a 1-off pipeline name.

File: cudaq/test/Transforms/aot_unwind_modes.qke [hunk 3]
```diff
@@ -0,0 +1,42 @@
+// ========================================================================== //
+// Copyright (c) 2026 NVIDIA Corporation & Affiliates.                        //
+//                                                                            //
+// This source code and the accompanying materials are made available under   //
+// the terms of the Apache License 2.0 which accompanies this distribution.   //
+// ========================================================================== //
+
+// RUN: cudaq-opt --dump-pass-pipeline --pass-pipeline='builtin.module(aot-prep-pipeline{unwind-mode=dataflow})' %s 2>&1 | FileCheck %s --check-prefix=DATAFLOW
```
Comment (schweitzpgi) [inline cudaq/test/Transforms/aot_unwind_modes.qke]: nit: could use `\` to wrap these long RUN lines.

File: cudaq/test/Transforms/unwind_by_dataflow.qke [hunk 4]
```diff
@@ -756,8 +756,10 @@ func.func @__nvqpp__mlirgen__function_9(%arg0: i1) -> i32 {
 // Structural check only
 // CHECK-LABEL:   func.func @__nvqpp__mlirgen__function_9(
 // CHECK:           %[[CONSTANT_0:.*]] = arith.constant 4 : i8
-// CHECK-DAG:       %[[ALLOCA_1:.*]] = cc.alloca i8
-// CHECK-DAG:       %[[ALLOCA_2:.*]] = cc.alloca i1
+// CHECK-DAG:       %[[NO_JUMP:.*]] = arith.constant 0 : i8
+// CHECK:           %[[ALLOCA_1:.*]] = cc.alloca i8
+// CHECK-NEXT:      cc.store %[[NO_JUMP]], %[[ALLOCA_1]] : !cc.ptr<i8>
```
Comment (schweitzpgi) [inline cudaq/test/Transforms/unwind_by_dataflow.qke]: As this is verified in tests above, I dropped these checks here and just checked the overall control-flow structure to reduce the size of the CHECK lines and avoid the redundancy. Was there a reason to add it back?

**Review by schweitzpgi — COMMENTED** [review]
Looking good. I think we should disentangle unwinding and inlining. I don't know why they are conflated in the pipeline today. Weirdness.

File: cudaq/lib/Optimizer/Transforms/UnwindByDataFlow.cpp
Comment (sacpis) [inline cudaq/lib/Optimizer/Transforms/UnwindByDataFlow.cpp:457]: Tried running the following code

```
import cudaq

cudaq.set_target("quake_fake")

@cudaq.kernel
def test(n: int) -> int:
    total = 0
    for i in range(3):
        if i == n:
            break
        total += 1
    return total

cudaq.run(test, 1, shots_count=1)
```

it fails with an error as
```
error: operand #0 does not dominate this use
Traceback (most recent call last):
  File "/usr/local/cudaq/cudaq/kernel/ast_bridge.py", line 6273, in compile_to_mlir
    cudaq_runtime.runPassManager(pm, bridge.module)
RuntimeError: pass pipeline failed
```

Seems like when a `cc.loop` has a step region, the pass moves the step region's body into a new guard cc.if but leaves the region's terminator behind. `%4` is now defined inside the `if`, so the IR fails the dominance verifier. The while-region case right below handles exactly this by yielding the value out of the guard. The step region case does not.

File: python/cudaq/kernel/utils.py [hunk 5]
```diff
@@ -45,6 +45,11 @@
     "entry-point kernels must discriminate first")
 
 
+def getAOTPassPipeline():
+    mode = cudaq_runtime.get_aot_unwind_mode()
+    return f"builtin.module(aot-prep-pipeline{{unwind-mode={mode}}})"
```
Comment (sacpis) [inline python/cudaq/kernel/utils.py]: The AOT pipeline string now depends on the active target (`get_aot_unwind_mode()`), but `_cached_qkeModule` is computed once, lazily, and is never invalidated when the target changes.

`PyKernel.qkeModule` has the same hasattr guard. Nothing is registered through `register_set_target_callback` to clear these.

So if I do something like this

```
# compiled with cfg
cudaq.set_target("qpp-cpu");  cudaq.run(k, 1, shots_count=1) 

# reuses the cfg module
cudaq.set_target("quake_fake"); cudaq.run(k, 1, shots_count=1)
```

The second call did not recompile. So a backend that asks for structured control flow silently receives CFG-lowered IR, which the mock server explicitly rejects and a real provider would reject too.

Let's key the cached AOT module on the compile target hash (`std::hash<PipelineConfig>` already includes `aotUnwindMode`), or drop the cache on target change?

File: cudaq/lib/Target/Yaml/TargetConfigYaml.cpp [hunk 6]
```diff
@@ -397,6 +401,14 @@ void MappingTraits<cudaq::config::BackendEndConfigEntry>::mapping(
   io.mapOptional("rules", info.ConditionalBuildConfigs);
 }
 
+std::string MappingTraits<cudaq::config::BackendEndConfigEntry>::validate(
+    IO &io, cudaq::config::BackendEndConfigEntry &info) {
+  if (info.AOTUnwindMode.empty() || info.AOTUnwindMode == "cfg" ||
+      info.AOTUnwindMode == "dataflow" || info.AOTUnwindMode == "none")
+    return {};
+  return "'unwind-mode' must be one of: cfg, dataflow, none.";
+}
```
Comment (sacpis) [inline cudaq/lib/Target/Yaml/TargetConfigYaml.cpp:413]: Seems like the new YAML unwind-mode check prints an error but does not reject anything? Let's have `parseTargetConfig` check `Input.error()` and throw, so the validator actually does something. We can do similar thing [here](https://github.com/NVIDIA/cuda-quantum/pull/5274/changes#diff-10cd832ac0036faf69d7489a209a01b2d0b9626a08feacb550383b2e81cf257dR290) as well.

File: unittests/nvqpp/backends/quake_backend/test_unwind_app.py [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+    total = 0
+    while i < 6:
+        if i == stop:
+            break
+        if i == skip:
+            i += 1
+            continue
+        total += i
+        i += 1
+    return total
```
Comment (sacpis) [inline unittests/nvqpp/backends/quake_backend/test_unwind_app.py:45]: Let's add for loop variants with break, continue and early return as well?

```
for i in range(3):
    if i == n:
        break
```

**Review by sacpis — COMMENTED** [review]
Overall LGTM. Thanks @1tnguyen. Left few comments.

File: cudaq/lib/Optimizer/Transforms/Pipelines.cpp
Comment (1tnguyen) [reply cudaq/lib/Optimizer/Transforms/Pipelines.cpp]: Renamed as `python-aot-pipeline` in https://github.com/NVIDIA/cuda-quantum/pull/5274/changes/2f952fb9a4a0b60ecc0b513a1071dd77b6af6d7a

File: cudaq/test/Transforms/aot_unwind_modes.qke (hunk 3 above)
Comment (1tnguyen) [reply cudaq/test/Transforms/aot_unwind_modes.qke]: Fixed all these long RUN lines in https://github.com/NVIDIA/cuda-quantum/pull/5274/changes/2f952fb9a4a0b60ecc0b513a1071dd77b6af6d7a

File: cudaq/test/Transforms/unwind_by_dataflow.qke (hunk 4 above)
Comment (1tnguyen) [reply cudaq/test/Transforms/unwind_by_dataflow.qke]: Removed those redundant `CHECK` lines in https://github.com/NVIDIA/cuda-quantum/pull/5274/changes/2f952fb9a4a0b60ecc0b513a1071dd77b6af6d7a

File: python/cudaq/kernel/utils.py (hunk 5 above)
Comment (1tnguyen) [reply python/cudaq/kernel/utils.py]: Good point! Fixed as suggested in https://github.com/NVIDIA/cuda-quantum/pull/5274/changes/77314c6514a11083e4a6fa0344d8c8567e041610

File: unittests/nvqpp/backends/quake_backend/test_unwind_app.py (hunk 7 above)
Comment (1tnguyen) [reply unittests/nvqpp/backends/quake_backend/test_unwind_app.py:45]: Yeah, this is an edge case that we're investigating. I'll add the test case when it is fixed.

File: cudaq/lib/Target/Yaml/TargetConfigYaml.cpp (hunk 6 above)
Comment (1tnguyen) [reply cudaq/lib/Target/Yaml/TargetConfigYaml.cpp:413]: Good point. Checked the error and throw in https://github.com/NVIDIA/cuda-quantum/pull/5274/changes/2f952fb9a4a0b60ecc0b513a1071dd77b6af6d7a
